editor/src/dev/mem.clj

```diff
@@ -0,0 +1,212 @@
+    If false, include references from weak references in the measurement.
+
+    :ignore-outer-class-reference
+    If true, ignores the outer class reference from non-static inner classes. In
+    practice this is only useful if the object provided to the measure function
+    is an instance of an inner class, and we wish to exclude outer class from
+    the measurement.
+
+    :ignored-classes
+    Seq of Class values to exclude from the measurement. The class hierarchy is
+    not considered, so you need to explicitly list the concrete classes to
+    exclude."
+  ^MemoryMeter [{:keys [debug
+                        ignore-known-singletons
+                        ignore-non-strong-references
+                        ignore-outer-class-reference
+                        ignored-classes]
+                 :as options
+                 :or {debug false
+                      ignore-known-singletons true
+                      ignore-non-strong-references true
+                      ignore-outer-class-reference false}}]
+  (validate-option-keys! options)
+  (let [strategy (make-strategy MemoryMeter/BEST)
+        ignored-class-filter (as-ignored-class-filter ignored-classes)
+        class-filter (make-class-filter ignore-known-singletons ignored-class-filter)
+        field-filter (make-field-filter ignore-known-singletons ignore-outer-class-reference ignore-non-strong-references ignored-class-filter)
+        listener-factory (make-listener-factory debug)]
+    (MemoryMeter. strategy class-filter field-filter listener-factory)))
+
+(defn size-text
+  "Return a human-readable string expressing the supplied byte-size using a
+  suitable SI unit."
+  ^String [^long byte-size]
+  (if (< byte-size 1000)
+    (str byte-size " B") ; Exact byte size.
+    (let [units "EPTGMK"]
+      (loop [size (/ byte-size 1000.0)
+             unit-index (dec (count units))]
+        (if (or (< size 1000.0)
+                (zero? unit-index))
+          (String/format Locale/ROOT "%.1f %cB" (to-array [size (.charAt units unit-index)]))
+          (recur (/ size 1000.0)
+                 (dec unit-index)))))))
+
+(defn measure
+  "Measure the memory usage of the supplied object. By default, the size is
+  returned as a human-readable string. If supplied, the map of options can be
+  used to control the return value and other details of the measurement.
+
+  Supported options:
+    :meter
+    Custom MemoryMeter. If not specified, use default.
+
+    :shallow
+    If true, count only the object header and its fields, don't follow object
+    references.
+
+    :bytes
+    If true, return a number of bytes instead of a string.
+
+    :debug
+    If true, print the object layout tree to stdout. Can also be set to a number
+    to limit the nesting level being printed.
+
+    :ignore-known-singletons
+    If false, include known singletons such as Enum, Class, ClassLoader and
+    AccessControlContext in the measurement.
+
+    :ignore-non-strong-references
```

Comment:
Nitpick: why not call it `:ignore-weak-references`?

Reply:
I think because `WeakReference` is specific type of weak reference, but there are several others like `PhantomReference` and `SoftReference` as well.

Reply:
Ah, it mirrors the naming and semantics from JAMM's Filters, got it!